doc/designs/ldap_pam_passthrough.md

```diff
@@ -0,0 +1,315 @@
+workaround.
+
+Currently PAM passthrough authentication basically works by accident
+and by working around the lack of direct handling of RADIUS in the
+password plugin. It would be better, and more secure, to deal directly
+with this in the IPA password plugin and not rely on side-effects.
+
+### IPA Framework plugin changes:
+
+For users with the RADIUS authentication indicator set:
+1. Require no userPassword and krbPrincipalKey 
+2. Require a radius proxy server set 
+3. Do not allow the RADIUS authentication indicator along with others
+   since the point of it is to outsource authentication.
+
+These will not be enforced retroactivly on upgrade since LDAP bind
+using RADIUS was not supported.
+
+### IPA password plugin changes:
+
+If the RADIUS authentication indicator is set on a user and the user
+has a userPassword or krbPrincipalKey and does not have a radius 
+proxy setting (ipatokenradiusconfiglink) then return
+LDAP_INVALID_CREDENTIALS.
+
+If the OTP authentication indicator is not set, in
+ipapwd_pre_bind_otp() return 0 if any authentication indicator is set.
+
+### The workflow
+
+ - RADIUS will be evaluated first.
+
+ - If RADIUS authtype is set:
+    - require no userPassword and krbPrincipalKey
+    - require radius proxy setting
+
+ - If RADIUS is an authentication indicator for a user on a BIND request
+     - If these conditions are not set then LOG_FATAL() and end the
+       authentication attempt by returning LDAP_INVALID_CREDENTIALS.
+     - Otherwise continue the authentication process.
+
+ - If RADIUS is not an authentication indicator then proceed with
+   authentication.
+
+ - If OTP is an authentication indictorn or the user
+     - Evaluate tokens using the existing workflow
+     - Otherwise fall back to PASSWORD authentication
+
+OTP checking is done in ipapwd_pre_bind_otp() which is called unless
+there is a sync request. Authentication indicator type handling needs to
+be better centralized here. There are two paths that can return different
+results (assuming otpreq = False).
+
+1. User has otp auth type and has no tokens it will return 0.
+2. User does not have otp auth type it will only return 0 if the
+   user has OTP_CONFIG_AUTH_TYPE_PASSWORD.
+
+This loophole needs to be closed. Perhaps change to return false if
+auth type is OTP_CONFIG_AUTH_TYPE_NONE. This would likely be more
+future-proof if more authentication indicators are added.
+
+If the OTP check doesn't return an error then the password will be
+authenticted, if there is one. This is the two-step first check
+OTP, then check the password.
+
+Since a plugin returning 0 will allow subsequent authentication
+for the case of RADIUS we need to enforce the no password and
+radius server requirements because with PAM passthrough enabled it
+becomes a defacto default method. In fact we may want to always ensure
+there is no RADIUS server defined if the RADIUS authentication
+indicator is not set.
+
```

Comment:
An additional thought is to simply bring the PAM pass-through method into ipa-pwd-extop, like we have in slapi-nis plugin. We can copy over the code from slapi-nis.

Reply:
done